### ISSUE TITLE
Cabal install should be smarter about /tmp mounted with noexec

### ISSUE THREAD

**Issue by yaccz** [haskell/cabal#2422]
```
<loads of build output>
Executing: dist/dist-sandbox-82ddaa05/build/Data/Yaml/Syck_hsc_make  >dist/dist-sandbox-82ddaa05/build/Data/Yaml/Syck.hs
running dist/dist-sandbox-82ddaa05/build/Data/Yaml/Syck_hsc_make failed (exit code 127)
command was: dist/dist-sandbox-82ddaa05/build/Data/Yaml/Syck_hsc_make  >dist/dist-sandbox-82ddaa05/build/Data/Yaml/Syck.hs
World file is already up to date.
cabal: Error: some packages failed to install:
...
```

It happened to me with 5 packages already and I worked around it by installing the lib as system package. Generally, this shows as <blah>, <x>, "_hsc_make failed (exit code 127)".

The main problem is my /tmp was mounted noexec. Now, you can parse /proc/self/mountinfo and check for that before trying to exec things from there.

Other problem is that I was not even aware it is using /tmp. The output looks like it's running in the $(pwd)/dist/. `cabal install <x>` doesn't show it. It shows with -v2 but just at the start, which never occured to me to look at since the error happens several pages down the build log.

And last problem is that it was incredibly difficult to debug. Thanks a lot to @dcoutts there. Big part of that is thanks to the fact that the build environment is dropped regardless of the build result. This (and other build failures) would probably be way easier to debug if the build environment was kept intact and cabal would also tell you the absolute path there, right below the error. At least that's what I'm used to when dealing with package build failures on gentoo or suse.

**Comment by 23Skidoo** [comment]
Related: #688, #677.

**Comment by BardurArantsson** [comment]
I appreciate security paranoia, but what's the use case here? Just throwing out a potential solution: Would being able to specify an "exec"-mounted directory instead of "/tmp" make you happier?

To the developers: Is there any particular reason for cabal to use a global /tmp directory[1] (instead of, say, a directory inside cabal's own build directory)?

[1] I realize that there are non-simple build systems that may do this, but let's assume Simple for argument's sake.

**Comment by yaccz** [comment]
Well, at least checking /proc/self/mountinfo for noexec and printing message more useful then ExitFaiure 127 would be nice.

**Comment by BardurArantsson** [comment]
(I don't mean to be antagonistic, I just couldn't find a good way to phrase this. I apologize in advance.)

Do you think a better error message would be sufficient?

FWIW, I'm not a fan of special-casing the OS unless it cannot be avoided by any reasonable means. For example, "/proc/..." is pretty Linux-specific, don't you think? What about other systems?

**Comment by yaccz** [comment]
For me yes. It was really difficult to track down.

As for other OSes, sorry I am not familiar with non-linux. Does such problem even exist there? Part of the problem here is that some distros ship with noexec as default so unaware users, like I was, of these distros will just go "what the hell is going on, I don't even".

**Comment by yaccz** [comment]
> FWIW, I'm not a fan of special-casing the OS.

You don't need to. Just oportunistically open the file and check it. It may not exist even on some linux.

**Comment by BardurArantsson** [comment]
> Part of the problem here is that some distros ship with noexec as default [...]

Hum. I think that was Ubuntu(?), but I thought they reverted that policy because quite a lot of stuff broke? (Doesn't change any releases that are "out there", obviously.)

I'm not sure what you mean by "open the file and check it"...? Concretely, how do you propose checking for "noexec"?

**Comment by kaorihinata** [comment]
If a C solution is acceptable, statfs(2) will do the job on Linux, *BSD and anything else that uses the BSD statfs(2) variant (like Darwin/Mac OS X). I've tested the following on Linux, FreeBSD and Mac OS X:

```
#include <sys/mount.h>
#ifdef __linux__
#include <sys/vfs.h>
#define __FLAG_NOEXEC MS_NOEXEC
#else
#define __FLAG_NOEXEC MNT_NOEXEC
#endif

int main() {
	struct statfs result;
	statfs("/tmp", &result);
	return result.f_flags & __FLAG_NOEXEC ? 0 : 1;
}
```

I'm sure there's a more appropriate set of ifdef's, but I'm on the clock right now and just ran into this issue.

**Comment by yaccz** [comment]
>  is pretty Linux-specific, don't you think? What about other systems?

I guess so. Is this a problem in other OS as well? Is there a standardized interface we could levarage to handle this issue? If not, is "we would have to implement an adapter pattern to make good UX for failure modes" a good argument for not fixing the issue?

>  Additionally, I'm of the opinion that making the tmp path configurable and making the error more clear would make more sense than trying to "be smart".

Which from my PoV is one possible implementation of "being smart". I don't have enough knowledge about neither the problem domain nor the other supported OS, which is why I just described the problem and requested to "make it better".

**Comment by yaccz** [comment]
Also easily fixable just by running `mount -o remount,noexec` but it is difficult to connect having just `ExitFailure 127` with /tmp noexec option. Especially when your primary linux distribution has /tmp with exec by default. Which is the primary problem here.

**Comment by kaorihinata** [comment]
> Is this a problem on other OS as well?

All Unix and Unix-like systems that honor the noexec flag in some capacity are affected.

> Which from my PoV is one possible implementation of "being smart".

Ahh, I apologize. I have always used "being smart" to mean "automating something harmless or simple so that the user doesn't have to worry about it". What I meant was that, since having `/tmp` mounted `noexec` is ostensibly a user decision, meeting the user half way by **not** hard coding the tmp path and making them aware of the issue would probably be the most sensible solution.

> Also easily fixable just by running `mount -o remount,noexec /tmp` ...

This is how I worked around the issue for my purposes, but I don't think it's safe to assume that the user will always have the rights required to remount `/tmp` with the `exec` flag.

**Comment by yaccz** [comment]
> This is how I worked around the issue for my purposes, but I don't think it's safe to assume that the user will always have the rights required to remount /tmp with the exec flag.

True. It appears the best possible solution is combination of informative message and ability to configure the /tmp location.

**Comment by SenorAgosto** [comment]
> For me yes. It was really difficult to track down.
> 
> As for other OSes, sorry I am not familiar with non-linux. Does such problem even exist there? Part of the problem here is that some distros ship with noexec as default so unaware users, like I was, of these distros will just go "what the hell is going on, I don't even".

OpenBSD mounts `/tmp` with `noexec` by default. Any `cabal install` trying to execute a process from that mount point fails on the default install.

**Comment by kerneis-anssi** [comment]
I think it's worth repeating here the workaround suggested in the linked blog post: 

> Fortunately, all of the packages I have come across that do this all respect the TMPDIR environment variable!
> 
> `alias cabal='env TMPDIR=/usr/local/cabal/build/ cabal'`
> 
> With this alias, you should be able to cabal without issue (so far pandoc, shellcheck and hakyll have all built fine)!
